### PR TITLE
Handle import-related errors in instantiate

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

https://github.com/facebookresearch/hydra/issues/799

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

To be discussed 

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

 https://github.com/facebookresearch/hydra/issues/799

### CODE REVIEW

**Comment by shagunsodhani** [comment]
If the current change looks good, I will add a test case.

**Comment by shagunsodhani** [comment]
Ohh good - I don't need to add new tests, just fixing the previous ones will do :)

File: hydra/_internal/utils.py [hunk 1]
```diff
@@ -516,8 +516,17 @@ def _locate(path: str) -> Union[type, Callable[..., Any]]:
...
-            raise ImportError(f"Could not locate '{path}'")
+            # Check if there was an import error
+            try:
+                import_module(mod)
+            except Exception as e:
+                raise ImportError(
+                    f"Encountered error: `{e}` when loading module '{path}'"
+                ) from e
+            else:
+                raise ImportError(f"Could not locate '{path}'")
```
Comment (shagunsodhani) [inline hydra/_internal/utils.py:529]: I dont think that this statement will ever be reached

Comment (omry) [inline hydra/_internal/utils.py:529]: Why is this needed? the stink is probably at callsite.

Comment (shagunsodhani) [reply hydra/_internal/utils.py:529]: `hasattr(obj, part)` is False if the imported file (ie `part`) has any errors (including importing an uninstalled module or syntax error).

Comment (omry) [reply hydra/_internal/utils.py:529]: We need to start working on improving the test coverage in Hydra.
There are some roadblocks like how to ensure examples coverage is properly accounted for.

Comment (omry) [reply hydra/_internal/utils.py:529]: Why do you need to intercept the exception here instead of in the callsite?

Comment (shagunsodhani) [reply hydra/_internal/utils.py:529]: You mean here: https://github.com/facebookresearch/hydra/blob/master/hydra/utils.py#L32 ?

Comment (omry) [reply hydra/_internal/utils.py:529]: yes. It's already catching exceptions. Can it use the message from the caught exception?

Comment (shagunsodhani) [reply hydra/_internal/utils.py:529]: Agreed that it is catching exceptions. But I still need `import_module(mod)` to generate the exception. So I will just generate the exception here, and pass it up all the way.

**Comment by shagunsodhani** [comment]
_I dont understand why the pre-commit does not fail locally_

**Comment by omry** [comment]
If this is wip please set it as draft.
otherwise please remove the wip/draft and request review.

File: hydra/_internal/utils.py [hunk 2]
```diff
@@ -516,8 +516,12 @@ def _locate(path: str) -> Union[type, Callable[..., Any]]:
     else:
         obj = builtins
     for part in parts[n:]:
+        mod = mod + "." + part
         if not hasattr(obj, part):
-            raise ImportError(f"Could not locate '{path}'")
+            try:
+                import_module(mod)
+            except ImportError as e:
+                raise e
```
Comment (omry) [inline hydra/_internal/utils.py:524]: So, if it's an import error, you catch it and raise it.
and if it's not an import error, you just let it fly.
see the problem?

Comment (shagunsodhani) [reply hydra/_internal/utils.py:524]: You are right. I (wrongly) assumed that `import_module` can only raise `ImportError` (https://github.com/python/cpython/blob/3.8/Lib/importlib/__init__.py#L109)

Comment (shagunsodhani) [reply hydra/_internal/utils.py:524]: ouch you are right. 

> Can you add an explicit test for a case where the instantiated class has an import error?

You have a test for this already; https://github.com/facebookresearch/hydra/pull/800/files#diff-01be86b771f92654b26fbe57dd65f9e7

**Comment by shagunsodhani** [comment]
> Can you add an explicit test for a case where the instantiated class has an import error?

You have a test for this already; https://github.com/facebookresearch/hydra/pull/800/files#diff-01be86b771f92654b26fbe57dd65f9e7

File: hydra/_internal/utils.py (hunk 2 above)
Comment (omry) [reply hydra/_internal/utils.py:524]: Even under this wrong assumption it makes no sense to catch and exception and then raise it without doing anything.

**Comment by omry** [comment]
> > Can you add an explicit test for a case where the instantiated class has an import error?
> 
> You have a test for this already; https://github.com/facebookresearch/hydra/pull/800/files#diff-01be86b771f92654b26fbe57dd65f9e7
> 
> I can still add an example that is closer to what OP raised.

If we have it already how did we have this bug?
this is a different error even though the exception is the same.

File: tests/test_utils.py [hunk 3]
```diff
@@ -231,6 +233,15 @@ def test_instantiate_adam() -> None:
     assert res == Adam(params=adam_params)
 
 
+def test_instantiate_with_missing_module() -> None:
+
+    with pytest.raises(
+        HydraException, match=re.escape(" No module named 'some_missing_module'")
```
Comment (omry) [inline tests/test_utils.py:239]: 1. test against the full exception message.
2. What does the exception look like in the console on this error?

Comment (shagunsodhani) [reply tests/test_utils.py:239]: 2. 
```
(hydra38) (mtrl) ➜  hydra_debug HYDRA_FULL_ERROR=1 python my_app.py
Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 522, in _locate
    import_module(mod)
  File "/private/home/sodhani/.conda/envs/hydra38/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/private/home/sodhani/hydra_debug/code/example_code.py", line 1, in <module>
    import pytorch_lightning.metrics.functional as plm
ModuleNotFoundError: No module named 'pytorch_lightning'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 33, in call
    type_or_callable = _locate(cls)
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 524, in _locate
    raise ImportError(
ImportError: Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 198, in run_and_report
    return func()
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 333, in <lambda>
    lambda: hydra.run(
  File "/private/home/sodhani/projects/hydra/hydra/_internal/hydra.py", line 105, in run
    return run_job(
  File "/private/home/sodhani/projects/hydra/hydra/core/utils.py", line 123, in run_job
    ret.return_value = task_function(task_cfg)
  File "my_app.py", line 7, in my_app
    python_class = hydra.utils.instantiate(cfg.code)
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 40, in call
    raise HydraException(f"Error calling '{cls}' : {e}") from e
hydra.errors.HydraException: Error calling 'code.example_code.sample_fucntion' : Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'
```

Comment (shagunsodhani) [reply tests/test_utils.py:239]: 2. 
```
Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 35, in call
    return _instantiate_class(type_or_callable, config, *args, **kwargs)
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 478, in _instantiate_class
    return clazz(*args, **final_kwargs)
  File "/private/home/sodhani/projects/hydra/tests/__init__.py", line 93, in __init__
    import some_missing_module  # type: ignore # noqa: F401
ModuleNotFoundError: No module named 'some_missing_module'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "tests/test_utils.py", line 252, in <module>
    utils.instantiate(ObjectConf(target="tests.ClassWithMissingModule"))
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 40, in call
    raise HydraException(f"Error calling '{cls}' : {e}") from e
hydra.errors.HydraException: Error calling 'tests.ClassWithMissingModule' : No module named 'some_missing_module'
```

Comment (shagunsodhani) [reply tests/test_utils.py:239]: 1. By full expecption message, you are referring to `hydra.errors.HydraException: Error calling 'tests.ClassWithMissingModule' : No module named 'some_missing_module'` correct ?

Comment (omry) [reply tests/test_utils.py:239]: 1. correct.
2. Is the above exception what the user sees normally or when when HYDRA_FULL_ERROR is used?

Comment (shagunsodhani) [reply tests/test_utils.py:239]: For this example, it doesnt make any difference. Adding the output for OP's code:
```
(hydra38) (mtrl) ➜  hydra_debug HYDRA_FULL_ERROR=0 python my_app.py
Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 522, in _locate
    import_module(mod)
  File "/private/home/sodhani/.conda/envs/hydra38/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/private/home/sodhani/hydra_debug/code/example_code.py", line 1, in <module>
    import pytorch_lightning.metrics.functional as plm
ModuleNotFoundError: No module named 'pytorch_lightning'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 33, in call
    type_or_callable = _locate(cls)
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 524, in _locate
    raise ImportError(
ImportError: Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "my_app.py", line 7, in my_app
    python_class = hydra.utils.instantiate(cfg.code)
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 40, in call
    raise HydraException(f"Error calling '{cls}' : {e}") from e
hydra.errors.HydraException: Error calling 'code.example_code.sample_fucntion' : Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

And

```
(hydra38) (mtrl) ➜  hydra_debug HYDRA_FULL_ERROR=1 python my_app.py
Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 522, in _locate
    import_module(mod)
  File "/private/home/sodhani/.conda/envs/hydra38/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/private/home/sodhani/hydra_debug/code/example_code.py", line 1, in <module>
    import pytorch_lightning.metrics.functional as plm
ModuleNotFoundError: No module named 'pytorch_lightning'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 33, in call
    type_or_callable = _locate(cls)
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 524, in _locate
    raise ImportError(
ImportError: Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 198, in run_and_report
    return func()
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 333, in <lambda>
    lambda: hydra.run(
  File "/private/home/sodhani/projects/hydra/hydra/_internal/hydra.py", line 105, in run
    return run_job(
  File "/private/home/sodhani/projects/hydra/hydra/core/utils.py", line 123, in run_job
    ret.return_value = task_function(task_cfg)
  File "my_app.py", line 7, in my_app
    python_class = hydra.utils.instantiate(cfg.code)
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 40, in call
    raise HydraException(f"Error calling '{cls}' : {e}") from e
hydra.errors.HydraException: Error calling 'code.example_code.sample_fucntion' : Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'
```

Comment (shagunsodhani) [reply tests/test_utils.py:239]: Hmm, even in this case, it doesnt make a difference so I think this is the standard output.

**Comment by omry** [comment]
Okay, I have a PR that is improving the error handling. I will land it once it passes CI.
try again with it and let me know.

**Comment by omry** [comment]
landed.

**Comment by shagunsodhani** [comment]
> Okay, I have a PR that is improving the error handling. I will land it once it passes CI.
> try again with it and let me know (#801).

```
(hydra38) (mtrl) ➜  hydra_debug python my_app.py
Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 526, in _locate
    import_module(mod)
  File "/private/home/sodhani/.conda/envs/hydra38/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/private/home/sodhani/hydra_debug/code/example_code.py", line 1, in <module>
    import pytorch_lightning.metrics.functional as plm
ModuleNotFoundError: No module named 'pytorch_lightning'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 33, in call
    type_or_callable = _locate(cls)
  File "/private/home/sodhani/projects/hydra/hydra/_internal/utils.py", line 528, in _locate
    raise ImportError(
ImportError: Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "my_app.py", line 7, in my_app
    python_class = hydra.utils.instantiate(cfg.code)
  File "/private/home/sodhani/projects/hydra/hydra/utils.py", line 40, in call
    raise HydraException(f"Error calling '{cls}' : {e}") from e
hydra.errors.HydraException: Error calling 'code.example_code.sample_fucntion' : Encountered error: `No module named 'pytorch_lightning'` when loading module 'code.example_code.sample_fucntion'

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

If this looks good, I can fix `test against the full exception message.` and push.

**Comment by omry** [comment]
This is way too verbose, can you put down your test code?

**Comment by omry** [comment]
we can have a followup diff improving the exception handling.
